sympy/solvers/solveset.py

```diff
@@ -1078,9 +1078,10 @@ def _solveset(f, symbol, domain, _check=False):
         elif isinstance(rhs_s, FiniteSet):
             for equation in [lhs - rhs for rhs in rhs_s]:
                 if equation == f:
+                    eq_expand = equation.expand()
                     if any(_has_rational_power(g, symbol)[0]
-                           for g in equation.args) or _has_rational_power(
-                           equation, symbol)[0]:
+                           for g in eq_expand.args) or _has_rational_power(
+                           eq_expand, symbol)[0]:
```

Comment:
Maybe rather than expanding here it would be better if `_has_rational_power` searched recursively through the expression.

Reply:
Expand can be slow for some equations so it should be avoided if possible.

Reply:
Actually, I think that even if  ``_has_rational_power ``recurses, finding a term with rational power will be very difficult. 
For example consider expressions  like 
`` expr = (x**Rational(2, 3)+ x**Rational(5, 3)+x**Rational(8, 3)+1)*(x**Rational(1, 3)) - 64``
For such cases, we can use simple heurestics, but consider many such terms with big powers, then it would be really difficult to find such a term

I think this problem is similar in nature to k-SAT problem which is undecidable.  Thus, it would be very hard to find an exact algorithm that can find such a term with rational powers. This is what I personally think, but I can be wrong.

Reply:
I don't understand why it would be hard in that case. Can we not just use `atoms(Pow)`?
```python
In [5]: expr.atoms(Pow)                                                                                                           
Out[5]: 
⎧3 ___   2/3   5/3   8/3⎫
⎨╲╱ x , x   , x   , x   ⎬
⎩                       ⎭
```

Reply:
Will this work for expressions like 
``(x**Rational(2, 3)+ x**Rational(5, 3)+x**Rational(8, 3)+1)**11  +.....``
What I mean is that the algorithm would  need to capture all possible terms, those which are evaluated after the we expand the power .
A specific example :
```
>>> ((sqrt(x)-4)**6).expand()
-24*x**(5/2) - 1280*x**(3/2) - 6144*sqrt(x) + x**3 + 240*x**2 + 3840*x + 4096
```
```
>>> f = (sqrt(x)-4)**6 + 24*x**Rational(5,2) + 1280*x**Rational(3,2) + 6144*x**Rational(1,2)
>>> atom = f.atoms(Pow)
>>> pprint(atom,use_unicode = False)
                               6
   ___   3/2   5/2  /  ___    \
{\/ x , x   , x   , \\/ x  - 4/ }
```

Reply:
I'm not sure I understand what the meaning of `_has_rational_power` is.

What is the expression that is being expanded to make it work in the two test cases here?

Reply:
I believe that _has_rational_power was implemented with the idea in mind that the terms of an expression will be passed nicely, that is they will be passed individually after all the processing is done.
In short, it looks for terms that have rational power. Then it returns a `` (bool, number)``,  bool answering, if the exprression has some rational power and the second value, is the denominator of that rational power (if it exists). I don't know what exactly is the use of the second returned value though.
For example:
```
>>> f = sqrt(x)
>>> _has_rational_power(f,x)
(True, 2)
>>> f = sqrt(x) -1
>>> _has_rational_power(f,x)
(False, 1)   # considers the whole expression as a single term, like (sqrt(x)-1)**1
```
Actually the expression  is  **``(sqrt(x)-4)**6``** . The first piece of code is just to highlight the representation of the function when expanded so that the values used in the second piece of code make sense.

Reply:
The fact that this checks equation and then equation.args looks fishy to me because it's sort of recursion but not quite. Why would we only want to recurse one level (especially without checking the whether expr is an Add, Mul etc.).

Can this be used:
```python
def _has_rational_power(expr, symbol):
    for power in expr.atoms(Pow, exp):
        base, exp = power.as_base_exp()
        if base.has(symbol) and exp.is_Rational and exp.q != 1:
            return True
```
Here `atoms` will recurse properly.

Reply:
This fail for ``(sqrt(x)-1)**2 + (sqrt(x)+1)**2``

Reply:
Does that actually cause a problem though? Would `_solve_as_radical` not work in that case?

Reply:
No, actually this code returns ``True`` even when there is no such term. 
I am sorry, I could not find  `` solve_as_radical`` . Did you mean ``solve_radical``?

Reply:
> Did you mean `_solve_radical`?

Yes, as in the function that is called on the line below the lines that we are commenting on. If `(sqrt(x)-1)**2 + (sqrt(x)+1)**2` were passed through to `_solve_radical` then would that not be able to solve it?

Reply:
Since this code returns True, ``_solve_radical`` will be called. 
I just tried this, but it ends up having infinite recursion. Most probably because it(the code in question) returns  ``True``, 
which leads to ``_solve_radical ``  which then ends up calling ``solveset`` and then back and forth between these.

Reply:
Is it necessary to check both `eq_expand` and `eq_expand.args` ? Would it be possible to run the check just for `Add.make_args(eq_expand)`?

I don't think that the check here is correct because it checks args without checking the type and it's not generally possible to know what the meaning of the args is without also checking the type. My guess is that args are checked because the object is assumed to be an Add.

If you try this you get infinite recursion error at this line:
```
solveset(sin(sqrt(1 + x)), x)
```

Reply:
@oscarbenjamin , I think you are right. There is no point in checking args individually (after expanding) and then having a separate check for the equation. 
``Add.make_args(eq_expand)``  should work assuming that ``expand()`` does expand every ``Pow`` in the expression.

This works
`` sovleset(sin(sqrt(x+1)), x, S.Reals)  ``